spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java

```diff
@@ -136,11 +140,54 @@ public void destroy() throws Exception {
 			configurationName = name;
 		}
 		Binder<T, ?, ?> binderInstance = getBinderInstance(configurationName);
-		Assert.state(GenericsUtils.getParameterType(binderInstance.getClass(), Binder.class, 0).isAssignableFrom(bindingTargetType),
+		Assert.state(
+				(binderInstance instanceof PollableConsumerBinder
+						&& checkCompatiblePollableBinder(binderInstance, bindingTargetType))
+					|| GenericsUtils.getParameterType(binderInstance.getClass(), Binder.class, 0)
+							.isAssignableFrom(bindingTargetType),
 				"The binder '" + configurationName + "' cannot bind a " + bindingTargetType.getName());
 		return binderInstance;
 	}
 
+	/*
+	 * Return the generic type of PollableSource to determine if it is appropriate
+	 * for the selected binder.
+	 * e.g. PollableMessageSource extends PollableSource<MessageHandler>
+	 * and  AbstractMessageChannelBinder
+	 *             implements PollableConsumerBinder<MessageHandler, C>
+	 * We're checking the the generic (MessageHandler) matches.
+	 */
+	@SuppressWarnings("rawtypes")
+	public boolean checkCompatiblePollableBinder(Binder binderInstance, Class<?> bindingTargetType) {
+		Class<?>[] binderInterfaces = ClassUtils.getAllInterfaces(binderInstance);
+		for (Class<?> intf : binderInterfaces) {
+			if (PollableConsumerBinder.class.isAssignableFrom(intf)) {
+				Class<?>[] targetInterfaces = ClassUtils.getAllInterfacesForClass(bindingTargetType);
+				Class<?> psType = findPollableSourceType(targetInterfaces);
+				if (psType != null) {
+					return GenericsUtils.getParameterType(binderInstance.getClass(), intf, 0)
+							.isAssignableFrom(psType);
+				}
+			}
+		}
+		return false;
+	}
+
+	private Class<?> findPollableSourceType(Class<?>[] targetInterfaces) {
```

Comment:
Wondering if we should consider adding this to `org.springframework.cloud.stream.reflection.GenericsUtils`?
```
public static Class<?> findType(Class<?>[] targetInterfaces, Class<?> type) {. . }
```